File: app/src/main/java/com/mmutert/freshfreezer/ui/itemlist/ItemListAdapter.java
package com.mmutert.freshfreezer.ui.itemlist;

import android.content.Context;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.AsyncListDiffer;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

import com.mmutert.freshfreezer.data.AmountUnit;
import com.mmutert.freshfreezer.data.FrozenItem;
import com.mmutert.freshfreezer.databinding.ListItemBinding;
import com.mmutert.freshfreezer.ui.ListSortingDialogFragment;
import com.mmutert.freshfreezer.util.SortingOption;
import com.mmutert.freshfreezer.viewmodel.FrozenItemViewModel;

import org.joda.time.LocalDate;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;

import java.text.NumberFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.Locale;


public class ItemListAdapter extends RecyclerView.Adapter<ItemListAdapter.ItemListAdapterViewHolder> implements
        ListSortingDialogFragment.ListSortingChangedListener {

    private final AsyncListDiffer<FrozenItem> mDiffer = new AsyncListDiffer<>(this, DIFF_CALLBACK);

    private final ListItemClickedCallback itemClickedCallback;
    private Context context;
    private FrozenItemViewModel mViewModel;


    public ItemListAdapter(
            final FrozenItemViewModel viewModel,
            ListItemClickedCallback itemClickedCallback,
            Context context) {
        this.mViewModel            = viewModel;
        this.itemClickedCallback   = itemClickedCallback;
        this.context               = context;
    }


    public void setItems(List<FrozenItem> items) {
        ArrayList<FrozenItem> newItems = new ArrayList<>(items);
        sortItems(newItems);
        mDiffer.submitList(newItems);
    }

    @NonNull
    @Override
    public ItemListAdapterViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {

        LayoutInflater inflater = LayoutInflater.from(parent.getContext());

        ListItemBinding binding = ListItemBinding.inflate(inflater, parent, false);

        return new ItemListAdapterViewHolder(binding);
    }

    @Override
    public void onBindViewHolder(@NonNull ItemListAdapterViewHolder holder, int position) {
        ListItemBinding binding = holder.binding;
        FrozenItem itemForPosition = getItemAtPosition(position);

        binding.setItem(itemForPosition);

        // TODO Format tv_amount
        AmountUnit unit = itemForPosition.getUnit();
        binding.tvAmountUnit.setText(context.getResources().getString(unit.getStringResId()));

        NumberFormat numberInstance = AmountUnit.getFormatterForUnit(itemForPosition.getUnit());
        String amountText = numberInstance.format(itemForPosition.getAmount());
        binding.tvAmount.setText(amountText);

        DateTimeFormatter formatter = DateTimeFormat.longDate().withLocale(Locale.getDefault());

        LocalDate bestBeforeDate = itemForPosition.getBestBeforeDate();
        String bestBeforeFormatted = formatter.print(bestBeforeDate);
        binding.tvBestBeforeDate.setText(bestBeforeFormatted);

        LocalDate frozenDate = itemForPosition.getFrozenAtDate();
        if(frozenDate != null) {
            binding.tvDateFrozen.setVisibility(View.VISIBLE);
            binding.tvFrozenDateTitle.setVisibility(View.VISIBLE);
            String frozenFormatted = formatter.print(frozenDate);
            binding.tvDateFrozen.setText(frozenFormatted);
        } else {
            binding.tvFrozenDateTitle.setVisibility(View.GONE);
            binding.tvDateFrozen.setVisibility(View.GONE);
        }
        binding.getRoot().setOnClickListener(v -> {
            itemClickedCallback.onClick(itemForPosition);
        });

        binding.listItemDeleteBackground.setVisibility(View.INVISIBLE);
        binding.listItemTakeBackground.setVisibility(View.INVISIBLE);
    }

    @Override
    public int getItemCount() {
        return mDiffer.getCurrentList().size();
    }

    public FrozenItem getItemAtPosition(int position) {
        return mDiffer.getCurrentList().get(position);
    }

    @Override
    public void listOptionClicked(
            final SortingOption selectedSortingOption, final SortingOption.SortingOrder sortingOrder) {
        mViewModel.setSortingOption(selectedSortingOption);
        mViewModel.setSortingOrder(sortingOrder);

        ArrayList<FrozenItem> frozenItems = new ArrayList<>(mDiffer.getCurrentList());
        sortItems(frozenItems);
        mDiffer.submitList(frozenItems);
    }

    private void sortItems(final List<FrozenItem> items) {
        if (items.size() > 0) {
            switch (mViewModel.getSortingOption()) {
                case DATE_CHANGED:
                    Collections.sort(items, (item1, item2) -> {
                        int result = item1.getLastChangedAtDate().compareTo(item2.getLastChangedAtDate());
                        if (mViewModel.getSortingOrder().equals(SortingOption.SortingOrder.ASCENDING)) {
                            return result;
                        } else {
                            return result * (-1);
                        }
                    });
                    break;
                case DATE_ADDED:
                    Collections.sort(items, (item1, item2) -> {
                        int result = item1.getItemCreationDate().compareTo(item2.getItemCreationDate());
                        if (mViewModel.getSortingOrder().equals(SortingOption.SortingOrder.ASCENDING)) {
                            return result;
                        } else {
                            return result * (-1);
                        }
                    });
                    break;
                case DATE_FROZEN_AT:
                    Collections.sort(items, (item1, item2) -> {
                        int result = item1.getFrozenAtDate().compareTo(item2.getFrozenAtDate());
                        if(mViewModel.getSortingOrder().equals(SortingOption.SortingOrder.ASCENDING)){
                            return result;
                        } else {
                            return result * (-1);
                        }
                    });
                    break;
                case DATE_BEST_BEFORE:
                    Collections.sort(items, (item1, item2) -> {
                        int result = item1.getBestBeforeDate().compareTo(item2.getBestBeforeDate());
                        if(mViewModel.getSortingOrder().equals(SortingOption.SortingOrder.ASCENDING)){
                            return result;
                        } else {
                            return result * (-1);
                        }
                    });
                    break;
                case NAME:
                    Collections.sort(items, (item1, item2) -> {
                        int result = item1.getName().compareTo(item2.getName());
                        if(mViewModel.getSortingOrder().equals(SortingOption.SortingOrder.ASCENDING)){
                            return result;
                        } else {
                            return result * (-1);
                        }
                    });
                    break;
            }
        }
    }

    /**
     * The Callback for the DiffUtil.
     */
    private static final DiffUtil.ItemCallback<FrozenItem> DIFF_CALLBACK
            = new DiffUtil.ItemCallback<FrozenItem>() {
        @Override
        public boolean areItemsTheSame(
                @NonNull FrozenItem oldFrozenItem, @NonNull FrozenItem newFrozenItem) {
            // FrozenItem properties may have changed if reloaded from the DB, but ID is fixed
            return oldFrozenItem.getId() == newFrozenItem.getId();
        }
        @Override
        public boolean areContentsTheSame(
                @NonNull FrozenItem oldFrozenItem, @NonNull FrozenItem newFrozenItem) {
            // NOTE: if you use equals, your object must properly override Object#equals()
            // Incorrectly returning false here will result in too many animations.
            return oldFrozenItem.equals(newFrozenItem);
        }
    };

    public List<FrozenItem> getItemList() {
        return mDiffer.getCurrentList();
    }


    /**
     * The view holder for the &
     */
    public class ItemListAdapterViewHolder extends RecyclerView.ViewHolder {

        public final ListItemBinding binding;

        public ItemListAdapterViewHolder(ListItemBinding binding) {
            super(binding.getRoot());
            this.binding = binding;
        }
    }
}

File: app/src/main/java/com/mmutert/freshfreezer/ui/itemlist/ItemListAdapter.kt
package com.mmutert.freshfreezer.ui.itemlist

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.AsyncListDiffer
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.mmutert.freshfreezer.data.AmountUnit.Companion.getFormatterForUnit
import com.mmutert.freshfreezer.data.StorageItem
import com.mmutert.freshfreezer.databinding.ItemOverviewItemBinding
import com.mmutert.freshfreezer.ui.dialogs.ListSortingDialogFragment.ListSortingChangedListener
import com.mmutert.freshfreezer.ui.itemlist.ItemListAdapter.ItemListAdapterViewHolder
import com.mmutert.freshfreezer.ui.itemlist.SortingOption.SortingOrder
import org.joda.time.format.DateTimeFormat
import java.util.*

class ItemListAdapter(
    private val mViewModel: ItemListViewModel,
    private val itemClickedCallback: ListItemClickedCallback,
    private val context: Context
) :
    RecyclerView.Adapter<ItemListAdapterViewHolder>(), ListSortingChangedListener {

    private val mDiffer = AsyncListDiffer(this, DiffCallBack())

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ItemListAdapterViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val binding = ItemOverviewItemBinding.inflate(inflater, parent, false)
        return ItemListAdapterViewHolder(binding)
    }

    override fun onBindViewHolder(holder: ItemListAdapterViewHolder, position: Int) {
        val binding = holder.binding
        val itemAtPosition = getItemAtPosition(position)

        val formatter = DateTimeFormat.longDate().withLocale(Locale.getDefault())
        val numberInstance = getFormatterForUnit(itemAtPosition.unit)

        binding.apply {
            item = itemAtPosition

            tvAmount.text = numberInstance.format(itemAtPosition.amount.toDouble())
            tvBestBeforeDate.text = formatter.print(itemAtPosition.bestBeforeDate)
            if (itemAtPosition.frozenAtDate != null) {
                tvDateFrozen.text = formatter.print(itemAtPosition.frozenAtDate)
            }
            root.setOnClickListener { itemClickedCallback.onClick(itemAtPosition) }
            listItemDeleteBackground.visibility = View.INVISIBLE
            listItemTakeBackground.visibility = View.INVISIBLE
        }
    }

    override fun getItemCount(): Int {
        return mDiffer.currentList.size
    }

    fun getItemAtPosition(position: Int): StorageItem {
        return mDiffer.currentList[position]
    }

    /**
     * Returns the current position of the given item as the index of the item in the list for the adapter
     * @param item The item to get the position for
     * @return The index of the item.
     */
    fun getPositionOfItem(item: StorageItem): Int {
        return mDiffer.currentList.indexOf(item)
    }

    override fun listOptionClicked(
        sortingOption: SortingOption,
        sortingOrder: SortingOrder
    ) {
        mViewModel.sortingOption = sortingOption
        mViewModel.sortingOrder = sortingOrder
        itemList = mDiffer.currentList
    }

    /**
     * Sorts the given list of items according to the sorting option and sorting order that are currently in the view model.
     * Sorting is done in-place.
     *
     * @param items  The list of items to sort.
     */
    private fun sortItems(items: MutableList<StorageItem>) {
        if (items.isNotEmpty()) {
            when (mViewModel.sortingOption) {
                SortingOption.DATE_CHANGED -> items.sortWith(Comparator { (_, _, _, _, _, _, _, lastChangedAtDate1), (_, _, _, _, _, _, _, lastChangedAtDate2) ->
                    val result = lastChangedAtDate1.compareTo(lastChangedAtDate2)
                    if (mViewModel.sortingOrder == SortingOrder.ASCENDING) {
                        return@Comparator result
                    } else {
                        return@Comparator result * -1
                    }
                })
                SortingOption.DATE_ADDED -> items.sortWith(Comparator { (_, _, _, _, _, _, itemCreationDate1), (_, _, _, _, _, _, itemCreationDate2) ->
                    val result = itemCreationDate1.compareTo(itemCreationDate2)
                    if (mViewModel.sortingOrder == SortingOrder.ASCENDING) {
                        return@Comparator result
                    } else {
                        return@Comparator result * -1
                    }
                })
                SortingOption.DATE_FROZEN_AT -> items.sortWith(Comparator { (_, _, _, _, frozenAtDate1), (_, _, _, _, frozenAtDate2) ->
                    var result = 0
                    if (frozenAtDate1 != null && frozenAtDate2 != null) {
                        result = frozenAtDate1.compareTo(frozenAtDate2)
                    } else if (frozenAtDate1 == null && frozenAtDate2 != null) {
                        result = -1
                    } else if (frozenAtDate1 != null && frozenAtDate2 == null) {
                        result = 1
                    }
                    if (mViewModel.sortingOrder == SortingOrder.ASCENDING) {
                        return@Comparator result
                    } else {
                        return@Comparator result * -1
                    }
                })
                SortingOption.DATE_BEST_BEFORE -> items.sortWith(Comparator { (_, _, _, _, _, bestBeforeDate1), (_, _, _, _, _, bestBeforeDate2) ->
                    val result = bestBeforeDate1.compareTo(bestBeforeDate2)
                    if (mViewModel.sortingOrder == SortingOrder.ASCENDING) {
                        return@Comparator result
                    } else {
                        return@Comparator result * -1
                    }
                })
                SortingOption.NAME -> items.sortWith(Comparator { (_, name), (_, name2) ->
                    val result =
                        name.lowercase(Locale.ROOT).compareTo(name2.lowercase(Locale.ROOT))
                    if (mViewModel.sortingOrder == SortingOrder.ASCENDING) {
                        return@Comparator result
                    } else {
                        return@Comparator result * -1
                    }
                })
            }
        }
    }

    var itemList: List<StorageItem>
        get() = mDiffer.currentList
        set(value) {
            val newItems = value.toMutableList()
            sortItems(newItems)
            mDiffer.submitList(newItems)
        }

    /**
     * The view holder
     */
    class ItemListAdapterViewHolder(val binding: ItemOverviewItemBinding) :
        RecyclerView.ViewHolder(binding.root)

    class DiffCallBack : DiffUtil.ItemCallback<StorageItem>() {
        override fun areItemsTheSame(
            oldStorageItem: StorageItem, newStorageItem: StorageItem
        ): Boolean {
            // FrozenItem properties may have changed if reloaded from the DB, but ID is fixed
            return oldStorageItem.id == newStorageItem.id
        }

        override fun areContentsTheSame(
            oldStorageItem: StorageItem, newStorageItem: StorageItem
        ): Boolean {
            // NOTE: if you use equals, your object must properly override Object#equals()
            // Incorrectly returning false here will result in too many animations.
            return oldStorageItem == newStorageItem
        }
    }

    companion object {
        val LOG_TAG = ItemListAdapter::class.simpleName
    }
}